lib/mix/test/mix/umbrella_test.exs

```diff
@@ -358,4 +358,30 @@ defmodule Mix.UmbrellaTest do
       assert_received {:mix_shell, :info, ["Compiled lib/foo.ex"]}
     end)
   end
+
+  test "apps cannot refer to themselves as a dep" do
+    in_fixture("umbrella_dep/deps/umbrella", fn ->
+      Mix.Project.in_project :umbrella, ".", fn _ ->
+        File.write! "apps/bar/mix.exs", """
+        defmodule Bar.Mixfile do
+          use Mix.Project
+
+          def project do
+            [app: :bar,
+             version: "0.1.0",
+             deps: [{:bar, in_umbrella: true}]]
+          end
+        end
+        """
+
+        assert_raise Mix.Error, "app bar refers to itself as a dependency", fn ->
+          Mix.Task.run("deps.get", ["--verbose"]) == [:ok, :ok]
+        end
+
+        assert_raise Mix.Error, "app bar refers to itself as a dependency", fn ->
+          Mix.Task.run("compile", ["--verbose"]) == [:ok, :ok]
+        end
```

Comment:
Do we need tests for both tasks? I think any task that ends up calling `Mix.Dep.Converger.topsort/1` is enough to test. Am I missing something? :)